### PR TITLE
Refactor RandomCoin trait

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
Addresses #206

### CODE REVIEW

File: prover/src/channel.rs [hunk 1]
```diff
@@ -137,18 +137,18 @@ where
...
     /// The positions are drawn from the public coin uniformly at random.
-    pub fn get_query_positions(&mut self) -> Vec<usize> {
+    pub fn get_query_positions(&mut self, nonce: u64) -> Vec<usize> {
         let num_queries = self.context.options().num_queries();
         let lde_domain_size = self.context.lde_domain_size();
         self.public_coin
-            .draw_integers(num_queries, lde_domain_size)
+            .draw_integers(num_queries, lde_domain_size, nonce)
             .expect("failed to draw query position")
     }
```
Comment (irakliyk) [inline prover/src/channel.rs:146]: I wonder if we actually need to pass in the `nonce` as a parameter here. It seems like we should be able to get this from `self` as `self.pow_nonce`. Or does this complicate something?

File: prover/src/channel.rs [hunk 2]
```diff
@@ -137,18 +137,18 @@ where
...
 
     /// Determines a nonce, which when hashed with the current seed of the public coin results
     /// in a new seed with the number of leading zeros equal to the grinding_factor specified
     /// in the proof options.
-    pub fn grind_query_seed(&mut self) {
+    pub fn grind_query_seed(&mut self) -> u64 {
```
Comment (irakliyk) [inline prover/src/channel.rs:151]: If the above comment applies, then we can probably leave this method as it was before.

File: prover/src/lib.rs [hunk 3]
```diff
@@ -414,10 +414,10 @@ pub trait Prover {
         let now = Instant::now();
 
         // apply proof-of-work to the query seed
-        channel.grind_query_seed();
+        let nonce = channel.grind_query_seed();
 
         // generate pseudo-random query positions
-        let query_positions = channel.get_query_positions();
+        let query_positions = channel.get_query_positions(nonce);
```
Comment (irakliyk) [inline prover/src/lib.rs:420]: Similar comment to above: we might be able to avoid making changes here.

File: crypto/src/random/mod.rs [hunk 4]
```diff
@@ -55,7 +47,8 @@ pub trait RandomCoin: Sync {
     /// PRNG.
     fn draw<E: FieldElement<BaseField = Self::BaseField>>(&mut self) -> Result<E, RandomCoinError>;
 
-    /// Returns a vector of unique integers selected from the range [0, domain_size).
+    /// Returns a vector of unique integers selected from the range [0, domain_size) after it reseeds
+    /// the coin with a nonce by setting the new seed to hash(`seed` || `nonce`).
```
Comment (irakliyk) [inline crypto/src/random/mod.rs:51]: nit: I wonder if we should not mention how exactly the reseeding happens as it concavely be done differently for different implementations of this trait.

File: crypto/src/random/default.rs [hunk 5]
```diff
@@ -208,7 +153,8 @@ impl<B: StarkField, H: ElementHasher<BaseField = B>> RandomCoin for DefaultRando
         Err(RandomCoinError::FailedToDrawFieldElement(1000))
     }
 
-    /// Returns a vector of unique integers selected from the range [0, domain_size).
+    /// Returns a vector of unique integers selected from the range [0, domain_size) after reseeding
+    /// the PRNG with the specified `nonce`.
```
Comment (irakliyk) [inline crypto/src/random/default.rs:157]: On the other hand, here I'd probably mention the exact methodology as this is a concrete implementation.

**Review by irakliyk — CHANGES_REQUESTED** [review]
Look good! Thank you! I left a few comments inline. Most of them are pretty small. The only more significant one is about using `nonce` from an instance variable rather than passing it via parameters. But if this significantly complicates FRI verifier interface, we can keep it as is.

File: prover/src/channel.rs (hunk 1 above)
Comment (Al-Kindi-0) [reply prover/src/channel.rs:146]: Makes total sense, we can indeed use `self.pow_nonce` to simplify things. I have done that now.

File: crypto/src/random/default.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply crypto/src/random/default.rs:157]: Makes sense.

**Review by irakliyk — APPROVED** [review]
All looks good! Thank you!

**Comment by irakliyk** [comment]
> The clippy error seems similar to the one [we hit on plonky2](https://github.com/mir-protocol/plonky2/pull/1236). Setting manually `CARGO_INCREMENTAL: 0` for the clippy run may do the trick.

Thank you! This seems to have fixed it.

**Pull request merged**